### DISCUSSION TITLE
Major performance refactoring

### DISCUSSION THREAD

**Discussion by czbar** [czbar/ChessForge#657] [Ideas]
Chess Forge was originally designed as a simple tool for learning (studying and memorizing) from a single variation tree. Viewing and editing of a single tree coupled with the Training mode was all that was envisaged.  The tree layout/editing and the way the data is loaded are pretty much unchanged.

Yet, the program has grown way beyond the original concept, boasting workbooks with chapters, and chapters with study trees, games and exercises, plus more.

Now, performance on large sets of data has to be considered unacceptable.  The main problems are:

1. Reading workbooks with large number of games takes a very long time with no indication to the user as to what is going on.
2. Building and displaying a large variation tree (say, 10,000+ moves) takes very long too.
3. Making moves within a large Study Tree can take excruciatingly long as the program repaints the view.
4. Large variation trees are difficult to grasp visually.

I intend to design and implement the following remediation:

Re 1: Upon opening a Workbook, the program will build the Chapters view and display the first view (which could be the Chapters view, or the last closed view in the previous session) before processing the trees (study/games/exercises). That should take a few seconds even for the largest of Workbooks. In the meantime, background threads will be started to process the trees. 

The GUI will remain responsive to the user. If the user clicks on a game that is still waiting in the queue to be processed, it will be immediately pushed to the front of the queue (i.e. a new processing thread will be started for it immediately).

Re 2: Displaying a variation tree must be completely revamped. A new design will be worked out.

Re 3: This will follow after item 2 has been implemented. When a move is made in the view, the changes made in the display should be minimized (currently a full repaint occurs). Ideally, the move should be shown on the chessboard immediately, the tree structure should be updated and the visual update of the tree view can be a little delayed.  

Re 4. In larger trees, smart expand/collapse will be implemented with appropriate visual hints for the user.

**Comment by akaKarthick** [comment]
I think your proposed remediation plan is a good one. It would address the performance issues that Chess Forge is currently experiencing with large sets of data. I hope you are able to implement it successfully!

**Comment by czbar** [comment]
Re 1:  in the private version 1.3.1, item 1 has been implemented and preliminarily tested with 20,000 games Workbook. The program reads the file first and extracts headers synchronously but then starts background processes to parse the games asynchronously. The GUI becomes responsive and the user can start working with the Workbook.  In the meantime, the progress is reported in the "Hints Box" under the main board.

Note that currently this only applies to opening a ChessForge Workbook. When opening a generic PGN, it still works the old way.

Also, some less impactful changes have been made e.g. when quickly going through a game (keeping the arrow key depressed) with Explorers on, the program does not call Lichess for opening stats on every move.
